src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+        if (!this.client.getLifecycleService().isRunning()) {
+            return Promise.reject(new ClientNotActiveError('Client is not active.'));
+        }
+
+        const connection = this.getConnectionFromAddress(address);
+        if (connection) {
+            return Promise.resolve(connection);
+        }
+
+        const addressKey = address.toString();
+        const pendingConnection = this.pendingConnections.get(addressKey);
+        if (pendingConnection) {
+            return pendingConnection.promise;
+        }
+
+        const connectionResolver: Promise.Resolver<ClientConnection> = DeferredPromise<ClientConnection>();
+        this.pendingConnections.set(addressKey, connectionResolver);
+
+        const processResponseCallback = (msg: ClientMessage) => {
+            this.client.getInvocationService().processResponse(msg);
+        };
+
+        this.translate(address)
+            .then((translatedAddress) => {
+                if (translatedAddress == null) {
+                    connectionResolver.reject(new RangeError(`Address Translator could not translate address ${address}`));
+                    return;
+                }
+
+                this.triggerConnect(translatedAddress)
+                    .then((socket) => {
+                        const clientConnection = new ClientConnection(this.client, translatedAddress, socket,
+                            this.connectionIdCounter++);
+                        return this.initiateCommunication(socket)
+                            .then(() => {
+                                return clientConnection.registerResponseCallback(processResponseCallback);
+                            })
+                            .then(() => {
+                                return this.authenticateOnCluster(clientConnection);
+                            }).then(() => {
+                                connectionResolver.resolve(clientConnection);
+                            });
+                    })
+                    .catch((error) => {
+                        connectionResolver.reject(error);
+                    })
+                    .finally(() => {
+                        this.pendingConnections.delete(addressKey);
+                    });
+            });
+
+        return connectionResolver.promise
+            .timeout(this.connectionTimeoutMillis, new HazelcastError(`Connection timed out to address ${address}.`))
+            .finally(() => {
+                this.pendingConnections.delete(addressKey);
+            });
+    }
+
+    public getRandomConnection(): ClientConnection {
+        if (this.isSmartRoutingEnabled) {
+            const member = this.loadBalancer.next();
+            if (member != null) {
+                const connection = this.getConnection(member.uuid);
+                if (connection != null) {
+                    return connection;
+                }
+            }
+        }
+
+        // TODO does the iterator.next() random enough ?
```

Comment:
`Map` objects iterate their elements in insertion order (see [this MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)), so `this.activeConnections.values().next()` doesn't give us any random at all. Although, I'm not sure if that's a problem for a unisocket client, which should have a single active connection only.